charms/worker/k8s/src/charm.py

```diff
@@ -271,12 +275,31 @@ def _create_cos_tokens(self):
     @on_error(
         WaitingStatus("Waiting for enable components"), InvalidResponseError, K8sdConnectionError
     )
-    def _enable_components(self):
+    def _configure_components(self):
         """Enable necessary components for the Kubernetes cluster."""
-        status.add(ops.MaintenanceStatus("Enabling DNS"))
-        self.api_manager.enable_component("dns", True)
+        dns_settings = self._dns_charm_integrated()
+        status.add(ops.MaintenanceStatus(f"Configuring DNS"))
+
+        if dns_settings:
+            self.api_manager.configure_component("dns", False)
+            self.api_manager.configure_dns(dns_settings.get("dns_domain"), dns_settings.get("dns_ip"))
+        else:
+            self.api_manager.configure_component("dns", True)
+
         status.add(ops.MaintenanceStatus("Enabling Network"))
-        self.api_manager.enable_component("network", True)
+        self.api_manager.configure_component("network", True)
+
+    def _dns_charm_integrated(self) -> dict:
```

Comment:
`Optional` is from the `typing` package

```suggestion
    def _dns_charm_integrated(self) -> Optional[dict]:
```